### ISSUE TITLE
Nerdfont is not being properly rendered

### ISSUE THREAD

**Issue by TonyDC** [hrkfdn/ncspot#1154]
**Describe the bug**
When using the `use_nerdfont = true` option, some visual glitches are present in the interface and the glyphs are not properly rendered.

**To Reproduce**
1. In the `config.toml` file, make sure that `use_nerdfont = true` is used
2. Start `ncspot` in the terminal. The visual glitches should start appearing
3. After the synchronization with Spotify is finished, when scrolling down, if you continue after the last entry in the page, the track length does not make sense
4. When jumping to `Podcasts` or `Browse` sections, the visual glitches are also present

**Expected behavior**
- The glyphs should be rendered in their proper place
- The track lengths should be correct
- The status bar should be rendered to the end

**Screenshots**

with `use_nerdfont = false` 
<img width="212" alt="image" src="https://user-images.githubusercontent.com/11735628/236715818-879429da-0f99-49c3-95d0-d46b22b371e5.png">

with `use_nerdfont = true` 
<img width="210" alt="image" src="https://user-images.githubusercontent.com/11735628/236715864-1993345d-b7b6-43d7-94b5-503c40df2fcf.png">

_Expected behavior_: the status bar should be rendered to the end

--- 

with `use_nerdfont = false`, after `ncspot` has synchronized with Spotify:
<img width="86" alt="image" src="https://user-images.githubusercontent.com/11735628/236715839-8cbf64a6-557a-45f1-b369-aa4b05312be3.png">

with `use_nerdfont = true`, after `ncspot` has synchronized with Spotify: 
<img width="321" alt="image" src="https://user-images.githubusercontent.com/11735628/236715917-eef69353-7a42-4030-a4e1-d3b3331abc6b.png">

with `use_nerdfont = true`, after `ncspot` has synchronized with Spotify and the user has scrolled past the last entry: 
<img width="179" alt="image" src="https://user-images.githubusercontent.com/11735628/236715984-1b098e4c-8910-425f-9e08-5606421056bc.png">

_Expected behavior_: the glyphs should be rendered in their proper place, the track lengths should be correct, and the status bar should be properly rendered

---

with `use_nerdfont = false`, after `ncspot` has synchronized with Spotify, in the `Podcasts` section:

<img width="64" alt="image" src="https://user-images.githubusercontent.com/11735628/236717394-f746d606-a90c-4989-98f7-bb8c9608e3b9.png">

with `use_nerdfont = true`, after `ncspot` has synchronized with Spotify, in the `Podcasts` section:

<img width="149" alt="image" src="https://user-images.githubusercontent.com/11735628/236716244-7678bd8c-037c-48e8-8eed-11b1d32ff689.png">

_Expected behavior_: it should be the same as when `use_nerdfont = false` is used

**System:**
 - OS: macOS Ventura 13.3.1
 - Terminal: iTerm2
 - Version: 0.13.2 
 - Installed from: brew

**Backtrace/Debug log**
Not applicable as `ncspot` hasn't crashed.

**Additional context**
- The font used was MesloMGS Nerd Font Mono. I also tried 3270 Nerd Font Mono and the results are the same. The latest version of the fonts were downloaded from <https://www.nerdfonts.com/font-downloads>
- These glitches started to appear after the upgrade to 0.13.2. The previous version didn't contain these issues. I suspect it might be related with #1141

**Comment by ThomasFrans** [comment]
I wanted to look into this but but I can't reproduce it. I still had version `0.13.0` installed, which did have problems with the newer Nerd Fonts. I now upgraded to `0.13.2` and there are no visual issues. I tried with both `ttf-nerd-fonts-symbols-mono` and `ttf-meslo-nerd`. Maybe it's related to the terminal.

**Comment by hrkfdn** [comment]
Same, I'm suspecting this may be terminal related..

**Comment by TonyDC** [comment]
@ThomasFrans @hrkfdn First of all, thank you both for you replies!

Indeed, if I run the version `0.13.2` with `use_nerdfont = true` in Ubuntu 22.04 using the default terminal emulator, the symbols are properly rendered (assuming the proper fonts are available) and the issues cannot be reproduced:
![image](https://github.com/hrkfdn/ncspot/assets/11735628/3c4a8628-a442-4464-8ab4-6cb5825a9949)

However, when I run on a mac, the interface starts falling apart. 

For testing purposes, I cloned the repo and executed `cargo run` in the following terminal emulators:
- iTerm2 (Build 3.4.19)
- Terminal (Version 2.13 (447)) - Note: this is the terminal emulator provided by macOS
- kitty (Version 0.28.1)

I also tried with `zsh` and `bash`. And, in all of them, I have the same issue. 

In addition, when using kitty, the progress bar is not being properly rendered (I just noticed this now because this was the first time I used kitty): 
<img width="647" alt="image" src="https://github.com/hrkfdn/ncspot/assets/11735628/120836ae-a99c-435e-b7c9-fede96582e6d">

Are you using macOS? If so, what terminal emulator are you using and how did you install the fonts?

**Comment by ThomasFrans** [comment]
> Are you using macOS? If so, what terminal emulator are you using and how did you install the fonts?

I tested on an Arch Linux install, so as expected, the results are the same as when you tested on Ubuntu.

**Comment by ucabgao** [comment]
I experienced the same glitches using ncspot 0.13.2 on macOS.

My setup is:

* Terminal emulator: alacritty 0.12.1
* Shell: fish 3.6.1
* OS: macOS Ventura 13.4

**Comment by hrkfdn** [comment]
Could anyone who's affected see if the same problem occurs with other terminal backends? `brew` for instance builds with `cursive/pancurses-backend`, but ncspot default is `termion`.

**Comment by TonyDC** [comment]
The interface is rendered as expected when the `termion` backend is used.


Here are the commands I used for each backend, as well as their respective outcomes:

## termion

```bash
git checkout v0.13.2
cargo run --locked --no-default-features --features portaudio_backend,termion_backend
```

<img width="218" alt="image" src="https://github.com/hrkfdn/ncspot/assets/11735628/92a4b3ec-45ae-4c91-9987-b5dc0e20ae34">

<img width="183" alt="image" src="https://github.com/hrkfdn/ncspot/assets/11735628/0c4ad7aa-bea5-44d3-988d-93436ab5a5f1">


## pancurses

```bash
git checkout v0.13.2
cargo run --locked --no-default-features --features portaudio_backend,pancurses_backend
```

<img width="216" alt="image" src="https://github.com/hrkfdn/ncspot/assets/11735628/ca263299-a924-48c9-a311-e9b6aefe9a2e">

<img width="362" alt="image" src="https://github.com/hrkfdn/ncspot/assets/11735628/acf7db28-033d-4d5a-8a49-8aa04d114307">

--- 

# Notes

- the font used was MesloMGS Nerd Font Mono

**Comment by TonyDC** [comment]
Also, I did a quick experiment using one of the examples provided by the cursive library (https://github.com/gyscos/cursive).

I added one of the glyphs to the dialog box:

```diff
diff --git a/cursive/examples/dialog.rs b/cursive/examples/dialog.rs
index c044205..695f0de 100644
--- a/cursive/examples/dialog.rs
+++ b/cursive/examples/dialog.rs
@@ -10,7 +10,7 @@ fn main() {
     // Creates a dialog with a single "Quit" button
     siv.add_layer(
         // Most views can be configured in a chainable way
-        Dialog::around(TextView::new("Hello Dialog!"))
+        Dialog::around(TextView::new("Hello Dialog! \u{f049d} "))
             .title("Cursive")
             .button("Foo", |_s| ())
             .button("Quit", |s| s.quit())

```

And then I experimented with various backends:

## ncurses

```bash
cargo run --example dialog --features ncurses-backend --no-default-features
```

<img width="300" alt="image" src="https://github.com/hrkfdn/ncspot/assets/11735628/f1a72d78-50f1-420c-8775-dd27637519d8">

## termion

```bash
cargo run --example dialog --features termion-backend --no-default-features
```

<img width="305" alt="image" src="https://github.com/hrkfdn/ncspot/assets/11735628/c0fb20ae-e0f1-4de7-b238-918968cdd38b">

## pancurses

```bash
cargo run --example dialog --features pancurses-backend --no-default-features
```

<img width="283" alt="image" src="https://github.com/hrkfdn/ncspot/assets/11735628/3b49583b-e017-4d9c-b5c1-f0195fcf1034">

Makes sense the results are the same as `ncurses`, since the calls are forwarded to [ncurses-rs](https://github.com/jeaye/ncurses-rs) in macOS (as described in the documentation: https://github.com/gyscos/cursive/wiki/Backends).

## crossterm

```bash
cargo run --example dialog --features crossterm-backend --no-default-features
```

<img width="279" alt="image" src="https://github.com/hrkfdn/ncspot/assets/11735628/add51732-107b-47eb-b899-b99d48419175">

---

It looks like it might be a problem with the UI library. What are your thoughts?

**Comment by hrkfdn** [comment]
Nothing we can do about this in ncspot I believe. Maybe a report to `cursive` makes sense? Could be the problem is not on their end, but they'll probably know better who should take a look at this.

**Comment by hrkfdn** [comment]
Of course, ncspot can just be built with termion. Are you referring to the Homebrew build? Someone would have to change the build options there as I don't maintain that.

**Comment by TonyDC** [comment]
Yes, I was referring to the Homebrew build, since that's the version I usually use on my mac.

**Comment by hrkfdn** [comment]
Closing this now as it would be solved via homebrew/homebrew-core#137010. Thanks everyone!
